dali/pipeline/executor/executor2/exec_node_task.cc

```diff
@@ -0,0 +1,507 @@
+    if (ws_->InputIsType<CPUBackend>(i)) {
+      auto inp = TaskInput<CPUBackend>(ti);
+      if (inp.event && inp.order != order)
+        events.insert(inp.event);
+      ws_->SetInput(i, inp.data);
+    } else {
+      assert(ws_->InputIsType<GPUBackend>(i));
+      auto inp = TaskInput<GPUBackend>(ti);
+      // If the output order of the operator is `host` then we don't wait for GPU
+      // inputs - they can't be accessed directly on host and the operator will
+      // have to issue some sort of synchronization if and when necessary.
+      // This optimization is essential to avoid oversynchronization
+      // when the operator needs to access the metadata only (e.g. getting the shape).
+      if (order.is_device() /*see comment above */ && inp.event && inp.order != order)
+        events.insert(inp.event);
+      ws_->SetInput(i, inp.data);
+    }
+  }
+
+  for (int i = 0; i < ws_->NumArgumentInput(); i++, ti++) {
+    auto &inp = TaskInput<CPUBackend>(ti);
+    if (inp.event)
+      events.insert(inp.event);
+    ws_->SetArgumentInput(i, inp.data);
+  }
+
+  if (task_->NumInputs() > ti) {  // Artificial error input
+    // TODO(michalz): Add early failure mechanism to `dali::tasking` that makes the successors
+    //                of a task that threw an exception fail. Currently the error disappears,
+    //                so we have to add an artificial input which propagates the exception.
+    task_->GetInputValue<void>(ti++);
+  }
+
+  for (auto e : events)
+    ws_->output_order().wait(e);
+
+  if (ws_->NumOutput()) {
+    if (ws_->NumInput() > 0) {
+      ws_->SetBatchSizes(ws_->GetInputBatchSize(0));
+    } else if (ws_->NumArgumentInput() > 0) {
+      ws_->SetBatchSizes(ws_->ArgumentInput(0).num_samples());
+    } else {
+      ws_->SetBatchSizes(ws_->GetIterationData()->default_batch_size);
+    }
+  }
+}
+
+AccessOrder OpTask::OutputConsumerOrder(int output_idx) const {
+  assert(static_cast<size_t>(output_idx) < node_->outputs.size());
+  // Return the common stream.
+  auto &consumers = node_->outputs[output_idx].consumers;
+  if (consumers.empty())
+    return {};  // definitely no consumer
+  AccessOrder order = consumers[0]->consumer->env.order;
+  for (size_t i = 1; i < consumers.size(); i++)
+    if (consumers[i]->consumer->env.order != order)
+      return {};
+  return order;
+}
+
+OpTask::OpTaskOutputs OpTask::GetWorkspaceOutputs() {
+  OpTaskOutputs ret;
+  int nout = ws_->NumOutput();
+  ret.reserve(nout);
+  auto order = ws_->output_order();
+  for (int o = 0; o < nout; o++) {
+    if (ws_->OutputIsType<CPUBackend>(o)) {
+      auto ptr = ws_->OutputPtr<CPUBackend>(o);
+      // If an input has a unique consumption order (all consumers use the same order),
+      // then we can simply move this buffer to a new order. The consumer stream
+      // will be properly syncrhonized in SetWorkspaceInputs.
+      // This is done to facilitate freeing of memory - if we're able to transfer the
```

Comment:
```suggestion
      // will be properly synchronized in SetWorkspaceInputs.
```

Reply:
The line moved again, there is a typo above in synchronized

Reply:
Fixed.